File: Ruby/Leetcode/two_sum.rb
def two_sum(nums, target)
  collect = {}
  nums.each_with_index do |key, idx|
    if collect[key].nil?
      collect[key] = { val: key, index: [idx] }
    else
      collect[key][:index] << idx
    end
  end

  collect.each do |key, elem|
    expected = target - key
    next unless collect[expected]

    if elem[:index].size > 1 && expected == key
      out = [elem[:index][0], collect[expected][:index][1]]
    else
      next if expected == key && collect[expected][:index].size == 1

      out = [elem[:index][0], collect[expected][:index][0]]
    end
    return out
  end
end

p two_sum([7, 9, 12], 21) # [1, 2]
p two_sum([7, 12, 9, 12], 24) # [1, 3]
p two_sum([2, 7, 11, 15], 9) # [0, 1]
p two_sum([3, 2, 4], 6) # [1, 2]

File: JS/leetcode/twosum.js
/**
* Given an array of integers, return indices of the two numbers such that they add up to a specific target.
*
* You may assume that each input would have exactly one solution, and you may not use the same element twice.
*
* Example:
*
* Given nums = [2, 7, 11, 15], target = 9,
*
* Because nums[0] + nums[1] = 2 + 7 = 9,
* return [0, 1].
*
*
* @param {number[]} nums
* @param {number} target
* @return {number[]}
*/


// Brute Force Approach
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */
const twoSum = function (nums, target) {
  let couple = undefined;
  for (idx = 0; idx < nums.length; idx ++) {
      for (jdx = idx + 1; jdx < nums.length; jdx ++) {
          if (nums[idx] + nums[jdx] == target) {
              couple = [idx, jdx];
              break;
          }
          if (couple) break;
      }
  }
  return couple;
};



var twoSum = function (nums, target) {
  const collection = indexCollector(nums);
  console.log(collection)
  let idx = 0
  let couple = [];
  let pairFound = false;

  while (!pairFound && idx < nums.length) {
    const value = nums[idx];
    const pair = target - value;

    if (pair == value && collection[value].length > 1) {

      couple = [collection[value][0], collection[value][1]];
      pairFound = true;

    } else if (collection.hasOwnProperty(pair)) {

      console.log(collection[value].length)
      couple = [collection[value][0], collection[pair][0]];
      pairFound = true;

  }

    idx++;
  }

  return couple;
};

function indexCollector(nums) {
  const collection = {};

  nums.forEach((num, idx) => {
    if (collection.hasOwnProperty(num)) {
      collection[num].push(idx);
    }
    else {
      collection[num] = [idx];
    }
  });
  return collection;
}


// console.log(twoSum([8, 7, 6, 15, 8], 16));
// console.log(twoSum([2, 7, 11, 15], 9));
console.log(twoSum([3, 2, 6], 6));
